src/common/Gateway.sol

```diff
@@ -275,8 +273,14 @@ contract Gateway is Auth, Recoverable, IGateway {
             }
         }
 
-        batcher = address(0);
         _endBatching(refund);
+        _batcher = address(0);
+    }
+
+    function lockCallback() external returns (address caller) {
+        require(address(_batcher) != address(0), CallbackIsLocked());
+        caller = _batcher;
+        _batcher = address(0);
     }
```

Comment:
It means:
- The callback that uses this can only be called once.
- The callback is called from the gateway under `withBatch`, otherwise `_batcher` will be 0.
- The callback is called from the same contract, because withBatch force `msg.sender`